src/ChainSafe.Gaming.Mud/MudExtensions.cs

```diff
@@ -1,16 +1,61 @@
-using System.Linq;
+using ChainSafe.Gaming.Mud.Storages;
+using ChainSafe.Gaming.Mud.Storages.InMemory;
+using ChainSafe.Gaming.Mud.Worlds;
 using ChainSafe.Gaming.Web3.Build;
 using ChainSafe.Gaming.Web3.Core.Nethereum;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
 
 namespace ChainSafe.Gaming.Mud
 {
     public static class MudExtensions
     {
+        /// <summary>
+        /// Configures and enables the use of MUD for the Web3 client that is being built.
+        /// </summary>
+        /// <param name="services">The Web3 service collection to configure.</param>
+        /// <param name="mudConfig">The MUD configuration.</param>
+        /// <returns>The same service collection that was passed in. This enables fluent style.</returns>
+        public static IWeb3ServiceCollection UseMud(this IWeb3ServiceCollection services, IMudConfig mudConfig)
+        {
+            services.AssertServiceNotBound<MudFacade>();
+            services.AssertConfigurationNotBound<IMudConfig>();
+
+            services.ConfigureMud(mudConfig);
+            services.UseMud();
+
+            return services;
+        }
+
+        /// <summary>
+        /// Configures the MUD services with the provided MUD configuration.
+        /// </summary>
+        /// <param name="services">The Web3 service collection to configure.</param>
+        /// <param name="mudConfig">The MUD configuration.</param>
+        /// <returns>The same service collection that was passed in. This enables fluent style.</returns>
+        public static IWeb3ServiceCollection ConfigureMud(this IWeb3ServiceCollection services, IMudConfig mudConfig)
+        {
+            services.Replace(ServiceDescriptor.Singleton(mudConfig));
+
+            return services;
+        }
+
+        /// <summary>
+        /// Enables the use of MUD for the Web3 client that is being built.
+        /// </summary>
+        /// <param name="services">The Web3 service collection to configure.</param>
+        /// <returns>The same service collection that was passed in. This enables fluent style.</returns>
         public static IWeb3ServiceCollection UseMud(this IWeb3ServiceCollection services)
         {
-            services.AddSingleton(typeof(MudFacade));
+            services.AssertServiceNotBound<MudFacade>();
+
+            services.AddSingleton<MudFacade>();
             services.AddSingleton<MudWorldFactory>();
+            services.AddSingleton<IMudStorageFactory, MudStorageFactory>();
+
+            // Storage strategies
+            services.AddTransient<InMemoryMudStorage>();
```

Comment:
Why is it a transient and also not the implementation of IMudStorage?

Reply:
If you look into `IInMemoryMudStorageConfig` you'll find a property with type `Type`. We then use this **concrete** type to request a service in service provider. Therefore we need the contract type to be the type of service itself and not an interface.

It's transient because we need to get a new instance each time we request a service of this type. This is because when we build a MUD world we configure this storage for the MUD world specifically.